File: backend/src/services/professionalService.js
import { ProfessionalRepository } from '../repositories/professionalRepository.js';

export class ProfessionalService {
  constructor(repo = new ProfessionalRepository()) {
    this.repo = repo;
  }

  async cadastrar({ nome, especialidade, telefone, email, servicos, ativo, diasTrabalho } = {}) {
    const n = String(nome || '').trim();
    const esp = String(especialidade || '').trim();
    const tel = String(telefone || '').replace(/\D/g, '');
    const em = String(email || '').trim().toLowerCase();

    if (!n) {
      const err = new Error('nome obrigatório');
      err.status = 400;
      throw err;
    }
    if (!esp) {
      const err = new Error('especialidade obrigatória');
      err.status = 400;
      throw err;
    }
    if (!tel || tel.length < 10) {
      const err = new Error('telefone obrigatório (mín. 10 dígitos)');
      err.status = 400;
      throw err;
    }
    if (!em || !/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(em)) {
      const err = new Error('email obrigatório inválido');
      err.status = 400;
      throw err;
    }

    let list = Array.isArray(servicos)
      ? servicos
      : String(servicos || '')
          .split(',')
          .map((s) => s.trim())
          .filter(Boolean);

    if (!list.length) list = [esp];

    const norm = [...new Set(list.map((s) => String(s).toUpperCase()))];
    return this.repo.create({
      nome: n,
      especialidade: esp,
      telefone: tel,
      email: em,
      servicos: norm,
      ativo: ativo !== false,
      diasTrabalho: Array.isArray(diasTrabalho) && diasTrabalho.length ? diasTrabalho.map(Number).filter((d) => d >= 0 && d <= 6) : [1, 2, 3, 4, 5],
    });
  }

  async listarAtivos() {
    return this.repo.listActive();
  }

  async listar() { return this.repo.listAll(); }

  async atualizarAtivo(id, ativo) { return this.repo.update(id, { ativo: Boolean(ativo) }); }

  async excluir(id) { return this.repo.delete(id); }

  async listarServicos() {
    const pros = await this.repo.listActive();
    const set = new Set();
    for (const p of pros) {
      (p.servicos || []).forEach((s) => set.add(String(s).toUpperCase()));
    }
    return [...set.values()].sort();
  }

  async profissionaisPorServico(servico) {
    const s = String(servico || '').trim().toUpperCase();
    if (!s) return [];
    const pros = await this.repo.listActive();
    return pros.filter((p) => (p.servicos || []).map(String).map((x) => x.toUpperCase()).includes(s));
  }
}

File: backend-java/src/main/java/com/clinica/bot/service/ProfessionalService.java
package com.clinica.bot.service;

import com.clinica.bot.domain.Profissional;
import com.clinica.bot.exception.ApiException;
import com.clinica.bot.repository.ProfessionalRepository;
import com.clinica.bot.util.CpfUtils;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.regex.Pattern;

@Service
@RequiredArgsConstructor
public class ProfessionalService {

    private static final Pattern EMAIL = Pattern.compile("^[^\\s@]+@[^\\s@]+\\.[^\\s@]+$");

    private final ProfessionalRepository repo;

    public Profissional cadastrar(Map<String, Object> body) {
        String nome = String.valueOf(body.getOrDefault("nome", "")).trim();
        String especialidade = String.valueOf(body.getOrDefault("especialidade", "")).trim();
        String telefone = CpfUtils.digitsOnly(String.valueOf(body.getOrDefault("telefone", "")));
        String email = String.valueOf(body.getOrDefault("email", "")).trim().toLowerCase();

        if (nome.isBlank()) throw new ApiException("nome obrigatório", 400);
        if (especialidade.isBlank()) throw new ApiException("especialidade obrigatória", 400);
        if (telefone.length() < 10) throw new ApiException("telefone obrigatório (mín. 10 dígitos)", 400);
        if (!EMAIL.matcher(email).matches()) throw new ApiException("email obrigatório inválido", 400);

        List<String> list;
        Object servicos = body.get("servicos");
        if (servicos instanceof List<?> l) {
            list = l.stream().map(String::valueOf).map(String::trim).filter(s -> !s.isBlank()).toList();
        } else {
            list = Arrays.stream(String.valueOf(servicos == null ? "" : servicos).split(","))
                    .map(String::trim).filter(s -> !s.isBlank()).toList();
        }
        if (list.isEmpty()) list = List.of(especialidade);
        List<String> norm = list.stream().map(s -> s.toUpperCase()).distinct().toList();

        List<Integer> diasTrabalho;
        Object dw = body.get("diasTrabalho");
        if (dw instanceof List<?> l && !l.isEmpty()) {
            diasTrabalho = l.stream().map(o -> ((Number) o).intValue()).filter(d -> d >= 0 && d <= 6).toList();
        } else {
            diasTrabalho = List.of(1, 2, 3, 4, 5);
        }

        Map<String, Object> data = new HashMap<>();
        data.put("nome", nome);
        data.put("especialidade", especialidade);
        data.put("telefone", telefone);
        data.put("email", email);
        data.put("servicos", norm);
        data.put("ativo", body.get("ativo") instanceof Boolean b ? b : true);
        data.put("diasTrabalho", diasTrabalho);
        return repo.create(data);
    }

    public List<Profissional> listar() { return repo.listAll(); }

    public Profissional atualizarAtivo(String id, Boolean ativo) {
        return repo.update(id, Map.of("ativo", Boolean.TRUE.equals(ativo)));
    }

    public Map<String, Object> excluir(String id) { return repo.delete(id); }

    public List<String> listarServicos() {
        Set<String> set = new TreeSet<>();
        for (Profissional p : repo.listActive()) {
            if (p.getServicos() != null) p.getServicos().forEach(s -> set.add(s.toUpperCase()));
        }
        return new ArrayList<>(set);
    }

    public List<Profissional> profissionaisPorServico(String servico) {
        String s = String.valueOf(servico == null ? "" : servico).trim().toUpperCase();
        if (s.isBlank()) return List.of();
        return repo.listActive().stream()
                .filter(p -> p.getServicos() != null && p.getServicos().stream().map(String::toUpperCase).anyMatch(x -> x.equals(s)))
                .toList();
    }
}
